File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    
    bool feasible(const vector<int>& piles,int speed,int hours){
        int h=0;
        for(int i:piles){
            int div=i/speed;
            h+=div;
            if(i%speed!=0)h++;
        }
        return h<=hours;
    }
    
    int minEatingSpeed(vector<int>& piles, int h) {
        
        int left=1,right=*max_element(piles.begin(),piles.end());
        while(left<right){
            int mid=left+(right-left)/2;
            if(feasible(piles,mid,h))
                right=mid;
            else left=mid+1;
        }
        return left;
        
        
//      int left=1;
//         int mid;
//         int right=1000000000;
//         while(left<=right){
//             mid=left+(right-left)/2;
//             if(can_possible(piles,mid,h)){
//                 right=mid-1;
                
//             }
//             else{
//                 left=mid+1;    
//             }
            
            
//         }
//         return left;
    }
//     bool can_possible(vector<int>&piles,int m,int h){
//         int hours=0;
//        for(int pile: piles){
//             int div=pile/m;
//             hours=hours+div;
//             if(pile%m!=0){
//                 hours++;
//             }
        
//         }
//         return hours<=h;
        
//     }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public boolean isFeasible(int[] piles,int mid,int h){
        int cnt=0;
        for(int i:piles){
            cnt+=i/mid;
            if(i%mid!=0)cnt++;
        }
        return cnt<=h;
   }
    public int minEatingSpeed(int[] piles, int h) {
        int left=1,right=1000000000;
        while(left<right){
            int mid=left+(right-left)/2;
            if(isFeasible(piles,mid,h))right=mid;
            else left=mid+1;
        }
        return left;
    }
}